src/common/snippets/src/lowered/pass/allocate_buffer_memory.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+
+#include "snippets/lowered/pass/allocate_buffer_memory.hpp"
+
+#include "snippets/lowered/pass/solve_buffer_memory.hpp"
+#include "snippets/lowered/pass/init_buffers.hpp"
+#include "snippets/lowered/pass/identify_buffers.hpp"
+#include "snippets/lowered/pass/define_buffer_clusters.hpp"
+#include "snippets/pass/tokenization.hpp"
+#include "snippets/itt.hpp"
+
+namespace ov {
+namespace snippets {
+namespace lowered {
+namespace pass {
+
+void AllocateBufferMemory::set_buffer_offset(const ExpressionPtr& buffer_expr, const size_t offset) {
+    // If Buffer has offset We set this offset in the connected MemoryAccess ops
+    // to correctly read and write data because all Buffers have the common data pointer on buffer scratchpad
+
+    const auto buffer = ov::as_type_ptr<op::Buffer>(buffer_expr->get_node());
+    OPENVINO_ASSERT(buffer, "Failed to set Buffer offset: AllocateBufferMemory expects Buffer op");
+    buffer->set_offset(static_cast<int64_t>(offset));
+
+    // Propagate to up: in Store. Buffer can have only one Store
+    {
+        if (buffer->is_intermediate_memory()) {
+            OPENVINO_ASSERT(buffer_expr->get_input_port_connectors().size() == 1, "Buffer with intermediate memory must have one parent");
+            const auto& parent_output = buffer_expr->get_input_port_connector(0)->get_source();
+            const auto& parent_expr = parent_output.get_expr();
+            const auto port = parent_output.get_index();
+            const auto& parent_node = parent_expr->get_node();
+            auto memory_access = ov::as_type_ptr<ov::snippets::op::MemoryAccess>(parent_node);
+            if (memory_access && memory_access->is_memory_access_output_port(port)) {
+                memory_access->set_output_offset(offset, port);
+            } else {
+                OPENVINO_THROW(
+                        "Buffer::set_offset() was called when Buffer didn't have the corresponding MemoryAccess op for offset propagation");
+            }
+        }
+    }
```

Comment:
Minor: why should we move this `if` in a separate block?
```suggestion
        if (buffer->is_intermediate_memory()) {
            OPENVINO_ASSERT(buffer_expr->get_input_port_connectors().size() == 1, "Buffer with intermediate memory must have one parent");
            const auto& parent_output = buffer_expr->get_input_port_connector(0)->get_source();
            const auto& parent_expr = parent_output.get_expr();
            const auto port = parent_output.get_index();
            const auto& parent_node = parent_expr->get_node();
            auto memory_access = ov::as_type_ptr<ov::snippets::op::MemoryAccess>(parent_node);
            if (memory_access && memory_access->is_memory_access_output_port(port)) {
                memory_access->set_output_offset(offset, port);
            } else {
                OPENVINO_THROW(
                        "Buffer::set_offset() was called when Buffer didn't have the corresponding MemoryAccess op for offset propagation");
            }
        }
```